### ISSUE TITLE
Interpolations containing HTML entities fail in IE11 when MutationObservers are active

### ISSUE THREAD

**Issue by zachsnow** [angular/angular.js#11781]
Text nodes that contain unicode / HTML entities are not properly interpolated in IE11 when there is a `MutationObserver` active.  Consider the following example.

```html
<!doctype html>
<html>
  <head>
    <script src="https://ajax.googleapis.com/ajax/libs/angularjs/1.4.0-rc.1/angular.min.js"></script>
    <script>
      window.addMutationObserver = function(){
        if(!window.MutationObserver){
          return;
        }
        new window.MutationObserver(function(){}).observe(document.body, {
          childList: true,
          subtree: true
        });
      };

      angular.module('example', []);
    </script>
  </head>
  <body onload="addMutationObserver()" ng-app="example">
    <span>{{ 'This works.' }}</span>
    <span>&mdash; {{ "This doesn't." }}</span>
  </body>
</html>
```

The output is:

```This works. — {{ "This doesn't." }}```

If you remove the `MutationObserver` the code works as expected. If you don't set `subtree: true` the code works as expected. If you remove the entity it also works as expected.  If instead you replace the entity with the actual character (—) it remains broken.  It's also broken in 1.3.

We recently started seeing this bug because Wordpress has some Emoji polyfills that use `MutationObserver`.  This issue breaks interpolation even when the Angular application is running in an iframe and the code that is attaching the `MutationObserver` is in the *parent* frame, which seems pretty crazy. Because of this we seek a workaround on the Angular side, as we can't change every site that might embed Angular in an iframe -- perhaps this is a bug in IE.

**Comment by zachsnow** [comment]
Using `ng-bind` in combination with additional nodes can work around the issue -- it's not always sufficient to wrap the interpolations in, say, `<span />` elements.

**Comment by lgalfaso** [comment]
This looks like a real issue, unlucky I do not have access to an IE11 so to diagnose this I will need some help

**Comment by gkalpak** [comment]
Took a brief look and here is what is going on:

```
// Given the following HTML:
<span>&mdash; {{ 'foo' }}</span>
var span = /* above <span> */

// In Chrome, IE11 (without MutationObserver):
span.childNodes = ["— foo"];   // All `textContent` is contained in 1 TextNode

// In IE11 (with MutationObserver):
span.childNodes = ["— ", "{{ '", "foo' }}"];   // The `textContent` is broken up into 3 TextNodes
```

So, in the former case, `$iterpolate` is called once with `— {{ 'foo' }}`, whereas in the latter it is called 3 times with `— `, `{{ '` and `foo' }}` as arguments respectively.

(I still don't know why this happens and if Angular can do something about it.)

--
@lgalfaso, I do have access to IE11; let me know if I can be of any help.

**Comment by lgalfaso** [comment]
@gkalpak is this also true if, after creating the `node` template out of the *string template*, we call `node.normalize()` ?

**Comment by lgalfaso** [comment]
@gkalpak or even simpler (even when not better), calling `node.normalize()` for every node that gets into `$compile`

**Comment by gkalpak** [comment]
@lgalfaso, will try it and get back to you (here or elsewhere).

BTW, I've been using this [CodePen][1] to reproduce the issue (should anyone be interested).

[1]: http://codepen.io/ExpertSystem/pen/OVVwqK?editors=101

**Comment by lgalfaso** [comment]
The first part is clear, this is an issue with IE11 as text elements must not be split. Now, if calling `normalize` would work as expected, then I think this would be a workaround that can be added to Angular. As @gkalpak found out, IE11 `normalize` is buggy too.
https://connect.microsoft.com/IE/feedback/details/809424

This implies that any solution will not end up being pretty. All that said, I think that there is nothing for angular to do to workaround the IE11 bug, but I will leave this issue open for people to comment and if someone wants to create a IE11 specific workaround, then to post it here.

**Comment by gkalpak** [comment]
IE11 + MutationObserver does some weird stuff after encountering the interpolation symbol (`{{`) and other special characters (e.g. `[[`) after an HTML entity.

One possible work-around is to use non-special characters as interpolation symbols. I am not sure what qualifies as "non-special", but alphabetic characters do. So, for example, using `aa` as the interolation start symbol (via `$interpolateProvider.startSymbol('aa')`) wil work.
(But it's such an ugly work-around :disappointed:)

**Comment by lgalfaso** [comment]
@gkalpak can you check if IE11 is happy with #11796 ?

**Comment by gkalpak** [comment]
@lgalfaso, IE11 is indeed happy with #11796.

**Comment by lgalfaso** [comment]
@petebacondarwin <https://github.com/petebacondarwin> WDYT

>

**Comment by petebacondarwin** [comment]
@lgalfaso: I commented on the PR

**Comment by shahata** [comment]
Spent a whole day nailing down this issue in one of our apps yesterday just to find out it is already reported and resolved :)

**Comment by gkalpak** [comment]
:laughing:

**Comment by petebacondarwin** [comment]
Got to keep on the bleeding edge @shahata !

**Comment by jshoudy11** [comment]
A similar error to this is coming up again in IE11 but only in CJK and Cyrillic languages. Must be related to how IE11 parses non-latin languages?

**Comment by jshoudy11** [comment]
It is splitting the interpolated string after the curly brackets. So it is becoming

<span>{{</span>
<span>interpolatedText}}</span>

**Comment by gkalpak** [comment]
@jshoudy11, do you have a demo ?

**Comment by jshoudy11** [comment]
This is actually a public facing issue on our website right now. 

- Go to https://console.cloud.google.com on IE11
- Click on the vertical "..." in the upper right hand corner
- Under "preferences" click "Language and regional formats"
- Choose one of the CJK languages
- Let the page reload and observe the "{{cardCtrl.project.name}}" in the top left card of the Dashboard

**Comment by Narretz** [comment]
You should open a new issue for that so it's easier to handle. We'll mark it as origin: google too.  And a small self contained demo is still valuable if you have one.

**Comment by gkalpak** [comment]
I couldn't reproduce it. It also doesn't make sense that changing an app setting would affect how the browser parses stuff. It was either a different problem or is related to the user's locale.

If you can reliably reproduce the problem, please open a new issue as @Narretz suggested (providing all relevant info, such as OS/version, browser/version, your locale etc).

**Comment by petebacondarwin** [comment]
@jshoudy11 are you using `ngMessageFormat` in the app that is causing problems?

**Comment by jshoudy11** [comment]
No. Not using ngMessageFormat. I'm trying to get a self contained demo that repros. I'll open a new issue if I get that.

**Comment by jshoudy11** [comment]
Also to get the repro follow the same instructions as before except look for "{{::link.text}} in the bottom right of the reloaded page
